src/test/fuzz/chainstate.cpp

```diff
@@ -0,0 +1,651 @@
+#include <util/thread.h>
+
+
+namespace {
+
+const BasicTestingSetup* g_setup;
+
+class KernelNotifications : public kernel::Notifications
+{
+public:
+    kernel::InterruptResult blockTip(SynchronizationState, CBlockIndex&) override { return {}; }
+    void headerTip(SynchronizationState, int64_t height, int64_t timestamp, bool presync) override {}
+    void progress(const bilingual_str& title, int progress_percent, bool resume_possible) override {}
+    void warning(const bilingual_str& warning) override {}
+    void flushError(const std::string& debug_message) override
+    {
+        assert(false);
+    }
+    void fatalError(const std::string& debug_message, const bilingual_str& user_message) override
+    {
+        assert(false);
+    }
+};
+
+auto g_notifications{KernelNotifications()};
+
+//! See net_processing.
+static const int MAX_HEADERS_RESULTS{2000};
+
+// We use a mapping from file path to buffer as a boutique in-memory file system. Note it's
+// fine because we only ever use unique pathnames for block files, but it may cause issues if
+// this assumption doesn't hold anymore. ("/a/b/blk0000.dat" and "/a/b/c/../blk0000.dat" point
+// to two different buffers.)
+// The use of a global does not prevent determinism (since the buffer from one run simply gets
+// overwritten in the next) but avoids a 128MB allocation per run. FIXME: is that really true?
+std::unordered_map<fs::path, std::vector<unsigned char>, std::hash<std::filesystem::path>> g_files;
+
+//! The initial block chain used to test the chainstate.
+std::vector<std::shared_ptr<CBlock>> g_initial_blockchain;
+
+void mock_filesystem_calls()
+{
+    fs::g_mock_create_dirs = [](const fs::path&) { return true; };
+    g_mock_check_disk_space = [](const fs::path&, uint64_t) { return true; };
+    fsbridge::g_mock_fopen = [&](const fs::path& file_path, const char* mode) {
+        // Get the file from the map. If it's not there insert it unless it's a file we aren't interested in.
+        const auto [data, size]{[&]{
+            const auto it = g_files.find(file_path);
+            if (it != g_files.end()) return std::make_pair(it->second.data(), it->second.size());
+            const auto file_name{PathToString(file_path.filename())};
+            // We shouldn't need to store anything else than the blk and rev files.
+            if (file_name.find("blk") == std::string::npos && file_name.find("rev") == std::string::npos) {
+                std::make_pair(nullptr, 0);
+            }
+            // NOTE: we do a single large alloc of the max possible file size, as there is no sane way to
+            // create shorter files but re-allocate when needed.
+            std::vector<unsigned char> buf(node::MAX_BLOCKFILE_SIZE);
+            const auto [it2, _]{g_files.insert({file_path, std::move(buf)})};
+            return std::make_pair(it2->second.data(), it2->second.size());
+        }()};
+        if (!data) return (FILE*)nullptr;
+        return fmemopen(data, size, mode);
```

Comment:
Have you checked if using `fmemopen` is better/faster than using a ram disk?

Reply:
I haven't. Can we use a ramdisk at oss-fuzz?

Reply:
Afaict oss-fuzz already uses a ram disk for its environments: https://github.com/google/clusterfuzz/blob/c461a961d8fb2afe47fb4af5eee3d1434a324a40/docker/base/setup_clusterfuzz.sh#L38 (i.e. `/tmp` is mounted in ram).

Reply:
https://joshua.hu/fuzzing-with-memfd-createfd-fmemopen-syscall-function

The author of this post found that a ram disk is slower🚀

I/O syscalls are avoided with fmempopen and everything simply happens in userland.

Reply:
Interesting. I was aware of `open_memstream` but not `memfd_create`. It's slower but could actually be helpful to get rid of the `128 MiB` allocs and make it possible to reindex.

Reply:
As discussed below i tried various ways of mocking the filesystem and ended up not mocking it at all: https://github.com/bitcoin/bitcoin/pull/29158#issuecomment-2214720674.